### PR TITLE
Infograph and InfographStar

### PR DESCRIPTION
**Pull request by tonydavis629**
# Pull Request Template

## Description

This PR adds another ModularTorchModel, InfoGraphModel, which is an unsupervised molecular representation learning model. The previous InfoGraphModel is renamed InfoGraphStar to match the literature's naming (InfoGraph*), which is a semi-supervised and supervised molecular property prediction learning model. Additional tests are written for both that test them with classification and regression datasets.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: docs/source/api_reference/models.rst [hunk 1]
```diff
@@ -399,6 +399,12 @@ InfoGraphModel
 .. autoclass:: deepchem.models.torch_models.InfoGraphModel
   :members:
 
+InfoGraphStarModel
+--------------
+
+.. autoclass:: deepchem.models.torch_models.InfoGraphStarModel
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:405]: Can you add both models to the model cheatsheet?

File: docs/source/api_reference/layers.rst [hunk 2]
```diff
@@ -166,6 +166,9 @@ Torch Layers
 .. autoclass:: deepchem.models.torch_models.InfoGraphEncoder
   :members:
 
+.. autoclass:: deepchem.models.torch_models.GINEncoder
```
Comment (rbharath) [inline docs/source/api_reference/layers.rst:169]: We should create a layers cheatsheet the way we have the models cheatsheet to track all the new layers. Could you take a crack at putting one together in a separate PR?

File: deepchem/models/torch_models/infograph.py [hunk 3]
```diff
@@ -6,6 +6,269 @@
...
+    Graph Information Network (GIN) encoder. This is a graph convolutional network that produces encoded representations for molecular graph inputs.
+
+    Parameters
+    ----------
+    num_features: int
+        The number of node features
+    embedding_dim: int
+        The dimension of the output embedding
+    num_gc_layers: int, optional (default 5)
+        The number of graph convolutional layers to use
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:27]: Can you add a usage example?

File: deepchem/models/torch_models/infograph.py [hunk 4]
```diff
@@ -6,6 +6,269 @@
...
+                                Linear(dim, dim))
+            else:
+                nn = Sequential(Linear(dim, dim), ReLU(), Linear(dim, dim))
+            conv = GINConv(nn)
+            bn = torch.nn.BatchNorm1d(dim)
+
+            self.convs.append(conv)
+            self.bns.append(bn)
+
+    def forward(self, data):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:58]: Can you add a docstring?

File: deepchem/models/torch_models/infograph.py [hunk 5]
```diff
@@ -6,6 +6,269 @@
...
+        self.init_emb()
+
+    def init_emb(self):
+        for m in self.modules():
+            if isinstance(m, nn.Linear):
+                torch.nn.init.xavier_uniform_(m.weight.data)
+                if m.bias is not None:
+                    m.bias.data.fill_(0.0)
+
+    def forward(self, data):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:119]: Can you add a docstring?

File: deepchem/models/torch_models/infograph.py [hunk 6]
```diff
@@ -6,6 +6,269 @@
...
+
+    def __init__(self, encoder, local_d, global_d, prior_d):
+        super().__init__()
+        self.encoder = encoder
+        self.local_d = local_d
+        self.global_d = global_d
+        self.prior_d = prior_d
+        self.init_emb()
+
+    def init_emb(self):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:112]: Can you add a docstring?

File: deepchem/models/torch_models/infograph.py [hunk 7]
```diff
@@ -6,6 +6,269 @@
...
+    The unsupervised training of InfoGraph involves two encoders: one that encodes the entire graph and another that encodes substructures of different sizes. The mutual information between the two encoder outputs is maximized using a contrastive loss function.
+    The model randomly samples pairs of graphs and substructures, and then maximizes their mutual information by minimizing their distance in a learned embedding space.
+
+    This can be used for downstream tasks such as graph classification and molecular property prediction.It is implemented as a ModularTorchModel in order to facilitate transfer learning.
+
+    References
+    ----------
+    .. [1] F.-Y. Sun, J. Hoffmann, V. Verma, and J. Tang, “InfoGraph: Unsupervised and Semi-supervised Graph-Level Representation Learning via Mutual Maximization.” arXiv, Jan. 17, 2020. http://arxiv.org/abs/1908.01000
+
+    Parameters
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:139]: Can you add a usage example?

File: deepchem/models/torch_models/infograph.py [hunk 8]
```diff
@@ -6,6 +6,269 @@
...
+        self.num_gc_layers = num_gc_layers
+        self.gamma = gamma
+        self.prior = prior
+        self.measure = measure
+        self.average_loss = average_loss
+        self.components = self.build_components()
+        self.model = self.build_model()
+        super().__init__(self.model, self.components, **kwargs)
+
+    def build_components(self) -> dict:
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:180]: Can you add a docstring explaining the components of the model? Model components should be documented for modular torch models in general

File: deepchem/models/torch_models/infograph.py [hunk 9]
```diff
@@ -6,6 +6,269 @@
...
+            'prior_d':
+                MultilayerPerceptron(self.embedding_dim,
+                                     1, (self.embedding_dim,),
+                                     activation_fn='sigmoid')
+        }
+
+    def build_model(self) -> nn.Module:
+        return InfoGraph(**self.components)
+
+    def loss_func(self, inputs, labels, weights):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:202]: Could you add a docstring?

Are you planning to shift this loss function to the deepchem losses in a future PR?

File: deepchem/models/torch_models/infograph.py [hunk 10]
```diff
@@ -6,6 +6,269 @@
...
+        if self.prior:
+            prior = torch.rand_like(y)
+            term_a = torch.log(self.components['prior_d'](prior)).mean()
+            term_b = torch.log(1.0 - self.components['prior_d'](y)).mean()
+            PRIOR = -(term_a + term_b) * self.gamma
+        else:
+            PRIOR = 0
+        return local_global_loss + PRIOR
+
+    def _local_global_loss(self, l_enc, g_enc, index):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:217]: This should be a function and not a method I think; it doesn't use self

File: deepchem/models/torch_models/infograph.py [hunk 11]
```diff
@@ -6,6 +6,269 @@
...
+        E_pos = get_positive_expectation(res * pos_mask, self.measure,
+                                         self.average_loss)
+        E_pos = (E_pos * pos_mask).sum() / pos_mask.sum()
+        E_neg = get_negative_expectation(res * neg_mask, self.measure,
+                                         self.average_loss)
+        E_neg = (E_neg * neg_mask).sum() / neg_mask.sum()
+
+        return E_neg - E_pos
+
+    def _prepare_batch(self, batch):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:253]: Can you add a docstring?

File: deepchem/models/torch_models/infograph.py [hunk 12]
```diff
@@ -51,36 +311,114 @@ def forward(self, data):
...
+        self.unsup_encoder = unsup_encoder
+        self.ff1 = ff1
+        self.ff2 = ff2
+        self.fc1 = fc1
+        self.fc2 = fc2
+        self.local_d = local_d
+        self.global_d = global_d
+        self.init_emb()
+
+    def init_emb(self):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:374]: Can you add a docstring?

File: deepchem/models/torch_models/infograph.py [hunk 13]
```diff
@@ -101,81 +439,129 @@ class InfoGraphModel(ModularTorchModel):
...
+            self.embedding_dim = embedding_dim * num_gc_layers
+            self.use_unsup_loss = True
+            self.num_gc_layers = num_gc_layers
+            self.separate_encoder = True
 
         self.components = self.build_components()
         self.model = self.build_model()
         super().__init__(self.model, self.components, **kwargs)
 
     def build_components(self):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:478]: Can you add a docstring and document the components?

**Review by rbharath — COMMENTED** [review]
The main feedback is that this PR is too large to review properly. Let's split this into 2 PRs. The first PR can rename the existing infograph to infographstar and the second PR should add the new infograph unsupervised model.

We need to make sure to document the components for each modular torch model. That should be available and easily readable from the docs for users as part of the official API. I have other documentation improvement requests added as well (let's add a layers cheatsheet and add the new models to the models cheatsheet).

File: deepchem/models/torch_models/infograph.py (hunk 10 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/infograph.py:217]: It does. It accesses self.device, self.measure and self.average_loss

File: deepchem/models/torch_models/infograph.py [hunk 14]
```diff
@@ -162,10 +177,183 @@ def forward(self, data):
...
+    def loss_func(self, inputs, labels, weights):
+        y, M = self.components['encoder'](inputs)
+        g_enc = self.components['global_d'](y)
+        l_enc = self.components['local_d'](M)
+        local_global_loss = self.localloss(l_enc, g_enc, inputs.graph_index)
+        if self.prior:
+            prior = torch.rand_like(y)
+            term_a = torch.log(self.components['prior_d'](prior)).mean()
+            term_b = torch.log(1.0 - self.components['prior_d'](y)).mean()
+            PRIOR = -(term_a + term_b) * self.gamma
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:336]: Let's not use the all caps style since we don't usually do that in deepchem code (so just `prior` is fine)

File: deepchem/models/torch_models/infograph.py [hunk 15]
```diff
@@ -162,10 +177,183 @@ def forward(self, data):
...
+                    m.bias.data.fill_(0.0)
+
+    def forward(self, data):
+        y, M = self.encoder(data)
+        g_enc = self.global_d(y)
+        l_enc = self.local_d(M)
+        return g_enc, l_enc
+
+
+class InfoGraphModel(ModularTorchModel):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:232]: Can you add this class to the docs? I think it's just added to the cheatsheet and not the docs

File: deepchem/models/torch_models/infograph.py [hunk 16]
```diff
@@ -162,10 +177,183 @@ def forward(self, data):
...
+        self.measure = measure
+        self.average_loss = average_loss
+        self.localloss = LocalMutualInformationLoss()._create_pytorch_loss(
+            measure, average_loss)
+        self.components = self.build_components()
+        self.model = self.build_model()
+        super().__init__(self.model, self.components, **kwargs)
+
+    def build_components(self) -> dict:
+        return {
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:306]: Can you add a table documenting the components and their names like you did for InfoGraphStar?

**Review by rbharath — COMMENTED** [review]
Almost good to go. A few missing things from the documentation but this should hopefully be ready to merge tomorrow

**Review by rbharath — APPROVED** [review]
LGTM, feel free to merge in once CI is clear
